### PR TITLE
Add A* arm trajectory planning

### PR DESCRIPTION
**Pull request by selym3**

### CODE REVIEW

**Comment by independence106** [comment]
NO NEGATIVE THOUGHTS

File: src/main/java/com/stuypulse/robot/commands/arm/ArmFollowTrajectory.java [hunk 1]
```diff
@@ -16,51 +19,52 @@ public class ArmFollowTrajectory extends CommandBase {
...
     }
 
-    public ArmFollowTrajectory(ArmTrajectory trajectory) {
-        this(() -> trajectory);
-    }
+    public ArmFollowTrajectory(Rotation2d finalShoulderAngle, Rotation2d finalWristAngle) {
+        arm = Arm.getInstance();
+
+        timer =  new StopWatch();
+        this.trajectory = AstarUtil.generateTrajectory(arm.getShoulderAngle().getDegrees(), arm.getWristAngle().getDegrees(), finalShoulderAngle.getDegrees(), finalWristAngle.getDegrees());
```
Comment (selym3) [inline src/main/java/com/stuypulse/robot/commands/arm/ArmFollowTrajectory.java:38]: don't couple the follow trajectory command with the astar file

Comment (selym3) [reply src/main/java/com/stuypulse/robot/commands/arm/ArmFollowTrajectory.java:38]: just remove this constructor

File: src/main/java/com/stuypulse/robot/util/AStar/AstarUtil.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+
+public class AstarUtil {
+
+    public static ArmTrajectory generateTrajectory(ArmState start, ArmState end) {
+        return generateTrajectory(
+            start.getShoulderState().getDegrees(),
+            start.getWristState().getDegrees(),
+            end.getShoulderState().getDegrees(),
+            end.getWristState().getDegrees()
```
Comment (selym3) [inline src/main/java/com/stuypulse/robot/util/AStar/AstarUtil.java:16]: take in constraint

File: src/main/java/com/stuypulse/robot/util/AStar/AstarUtil.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        return trajectory;
+    }
+
+    public static Constraint generateHorizontalLineConstraint(int startX, int startY, int length) {
+        return (s, w) -> {
+            return (startX <= s && s <= startX + length) && w == startY;
+        };
+    }
+
+    public static Constraint generateVerticalLineConstraint(int startX, int startY, int length) {
```
Comment (selym3) [inline src/main/java/com/stuypulse/robot/util/AStar/AstarUtil.java:50]: maybe just call this get.....Constraint

File: src/main/java/com/stuypulse/robot/util/AStar/AstarUtil.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    public static ArmTrajectory generateTrajectory(ArmState start, ArmState end) {
+        return generateTrajectory(
+            start.getShoulderState().getDegrees(),
+            start.getWristState().getDegrees(),
+            end.getShoulderState().getDegrees(),
+            end.getWristState().getDegrees()
+        );
+    }
+
+    public static ArmTrajectory generateTrajectory(double initialShoulderAngle, double initialWristAngle, double finalShoulderAngle, double finalWristAngle) {
```
Comment (selym3) [inline src/main/java/com/stuypulse/robot/util/AStar/AstarUtil.java:20]: take constraint in as parameter

File: src/main/java/com/stuypulse/robot/util/DoubleJointedArmSim.java [hunk 5]
```diff
@@ -20,9 +20,9 @@ public DoubleJointedArmSim(
             shoulderMomentOfInertia + wristMomentOfIntertia, 
             shoulderArmLengthMeters + wristArmLengthMeters, 
             shoulderMinAngleRads, shoulderMaxAngleRads,
-             true);
+             false);
 
-        wristSim = new SingleJointedArmSim(wristDCMotorGearbox, wristGearing, wristMomentOfIntertia, wristArmLengthMeters, wristMinAngleRads, wristMaxAngleRads, true);
+        wristSim = new SingleJointedArmSim(wristDCMotorGearbox, wristGearing, wristMomentOfIntertia, wristArmLengthMeters, wristMinAngleRads, wristMaxAngleRads, false);
```
Comment (selym3) [inline src/main/java/com/stuypulse/robot/util/DoubleJointedArmSim.java:25]: what?

File: src/main/java/com/stuypulse/robot/util/DoubleJointedArmSim.java [hunk 6]
```diff
@@ -20,9 +20,9 @@ public DoubleJointedArmSim(
             shoulderMomentOfInertia + wristMomentOfIntertia, 
             shoulderArmLengthMeters + wristArmLengthMeters, 
             shoulderMinAngleRads, shoulderMaxAngleRads,
-             true);
+             false);
```
Comment (selym3) [inline src/main/java/com/stuypulse/robot/util/DoubleJointedArmSim.java:23]: what?

File: src/main/java/com/stuypulse/robot/subsystems/Manager.java [hunk 7]
```diff
@@ -97,6 +102,27 @@ public ArmTrajectory getIntakeTrajectory() {
...
+        };
+
+        ArmTrajectory intakeTrajectory = new ArmTrajectory();
+
+        for (int i = 0; i < states.length - 1; ++i) {
+            intakeTrajectory.append(AstarUtil.generateTrajectory(
+                states[i], states[i+1]));
+        }
+        
+        return intakeSide == IntakeSide.FRONT ? intakeTrajectory : intakeTrajectory.flipped();
```
Comment (selym3) [inline src/main/java/com/stuypulse/robot/subsystems/Manager.java:123]: remove sample jason

File: src/main/java/com/stuypulse/robot/subsystems/arm/PerfectSimArm.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.math.util.Units;
+
+import edu.wpi.first.wpilibj.smartdashboard.FieldObject2d;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+public class PerfectSimArm extends Arm {
```
Comment (selym3) [inline src/main/java/com/stuypulse/robot/subsystems/arm/PerfectSimArm.java:21]: Pull from main (should do this anyway) and DELETE this file

File: src/main/java/com/stuypulse/robot/subsystems/Manager.java (hunk 7 above)
Comment (selym3) [reply src/main/java/com/stuypulse/robot/subsystems/Manager.java:123]: *Remove sample trajectory, Jason.

**Review by selym3 — COMMENTED** [review]
Looks alright. If I get some more proof this works I can merge it

**Comment by independence106** [comment]
All requested changes were made

**Pull request closed without merging**